third_party/move/move-vm/runtime/src/storage/loader/native.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    storage::loader::traits::StructDefinitionLoader, ModuleStorage, RuntimeEnvironment,
+    WithRuntimeEnvironment,
+};
+use move_binary_format::errors::PartialVMResult;
+use move_vm_types::{
+    gas::{GasMeter, ModuleTraversalContext},
+    loaded_data::{runtime_types::StructType, struct_name_indexing::StructNameIndex},
+};
+use std::sync::Arc;
+
+/// Loader implementation used for natives and function value serialization. Prior to lazy loading,
+/// natives were not metering gas for modules. With lazy loading:
+///   1. For metering layout construction in native context. Because natives do not have access to
+///      gas meter trait, lazy loader currently pre-loads the transitive closure for that native
+///      and its type arguments. Hence, all layouts in native context must be fully metered.
+///   2. For function value serialization. Function values may capture arguments, and if stored to
+///      the global storage, layouts are constructed for these captured arguments. Note that we
+///      expect module used for layouts to be already loaded (because construction of captured
+///      arguments implies their layouts are loaded as well).
+// TODO(lazy-loading):
+//   It seems that (2) does not hold with enums though. If captured argument comes from the stack,
+//   some A::V2 for example, modules for A::V1 will not be loaded, but they will be accessed during
+//   layout construction... It seems that what "saves" us here is that we will check the depth of
```

Comment:
I believe you confirmed this?

Question about (1). Is it conceivable that we could give gas meter access to natives and avoid the eager pre-loading by lazy loader? Then we could also document that TODO here.

Reply:
Yes, but this requires refactoring of gas meter to make it passable as `dyn` or natives to make them generic (which are both big, and I think in plans for Victor but I am not sure we can rely on a specific timeline). I am thinking that actually eager pre-loading is not too bad because layouts will be used. What we can do, actually, is allow native to implement traits, e.g. `UsesLayout` or `UsesAnnotatedLayout` so we can prefetch in a smart way, wdyt?

Reply:
Update: PR above this one uses a native function to return control to the VM to charge gas.